prov/util/src/util_coll.c

```diff
@@ -249,7 +239,7 @@ static int util_coll_sched_recv(struct util_coll_mc *coll_mc, int src,
 	xfer_item->hdr.type = UTIL_COLL_RECV;
 	xfer_item->hdr.is_barrier = is_barrier;
 
-	xfer_item->hdr.tag = (recv_tag << 32) | (tag & 0xffffffff);
+	xfer_item->hdr.tag = (src_rank << 32) | (tag & (0xffffffff | OFI_TAG_COLL));
```

Comment:
nit: I would restructure this to avoid making it look like OFI_TAG_COLL is associated with the lower 32-bits.  E.g.
```
OFI_TAG_COLL | (src_rank << 32) | (tag & 0xffffffff)
```
I would even look at removing the manipulation of the tag here completely, so that we just do
```
hdr.tag = tag  /* or htonl(tag) */
```
and fold in the src_rank value when we get the tag.

Reply:
the intent was to just mask the 31 bits that are the 'source' rank of the message rather than forcing the 'collective bit' to be set.  In that case, `OFI_TAG_COLL` is actually associated with the lower 32-bits in that together (bitwise OR) they form the rank mask `0x80000000FFFFFFFF`.  A bitwise AND with this mask would 'zero out' the rank part of the tag, but not modify the 'collective flag', cid, or seq parts of the tag.

Reply:
I'm confused, unless 'tag' has the upper bit set, this clears the OFI_TAG_COLL flag.  If you just want to clear the source rank bits, I would do that more explicitly, then set them.  E.g.
```
tag = tag & ~OFI_TAG_SOURCE_BITS;
tag |= (src_rank << OFI_TAG_SOURCE_SHIFT)
```
which you could put into a static inline call if it's needed elsewhere.